app/src/main/kotlin/compiler/intermediate/generators/SysV64CallingConvention.kt

```diff
@@ -7,41 +7,59 @@ import compiler.intermediate.Register
 
 val argPositionToRegister = listOf(Register.RDI, Register.RSI, Register.RDX, Register.RCX, Register.R8, Register.R9)
 val callerSavedRegisters = listOf(Register.RAX, Register.RCX, Register.RDX, Register.RSI, Register.RDI, Register.R8, Register.R9, Register.R10, Register.R11)
+const val stackAlignmentInBytes = 16
 
 object SysV64CallingConvention {
     fun genCall(targetFunction: IFTNode.MemoryLabel, args: List<IFTNode>, returnsValue: Boolean): FunctionDetailsGenerator.FunctionCallIntermediateForm {
-        val cfgBuilder = ControlFlowGraphBuilder()
+        val moveArgsCFGBuilder = ControlFlowGraphBuilder()
         val usedRegisters = mutableListOf<Register>()
 
         // First, move arguments to appropriate registers (or push to stack) according to call convention.
         for ((arg, register) in args zip argPositionToRegister) {
             val node = IFTNode.RegisterWrite(register, arg)
-            cfgBuilder.addLinksFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
+            moveArgsCFGBuilder.addLinksFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
             usedRegisters.add(register)
         }
 
-        var numberOfArgsPushedToStack = 0
+        // Before pushing args stack will be aligned, so code adds links to separate CFGBuilder and merges them later
+        var stackMovement = 0L
         for (arg in args.drop(argPositionToRegister.size).reversed()) {
-            val node = IFTNode.StackPush(arg)
-            cfgBuilder.addLinksFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
-            numberOfArgsPushedToStack += 1
+            moveArgsCFGBuilder.addLinksFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, IFTNode.StackPush(arg))
+            stackMovement += memoryUnitSize.toLong()
         }
 
+        val cfgBuilder = ControlFlowGraphBuilder()
+        // Align stack
+        if (stackMovement % stackAlignmentInBytes != 0L) {
+            val toAdd = stackAlignmentInBytes - stackMovement % stackAlignmentInBytes
+            cfgBuilder.addLinksFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IFTNode.Subtract(
+                    IFTNode.RegisterRead(Register.RSP),
+                    IFTNode.Const(toAdd)
+                )
```

Comment:
You need to actually save the result to RSP (`IFTNode.RegisterWrite`).

Reply:
Great catch, fixed